### PR TITLE
SI-8117 Fix bug when mixing well-positioned named and positional args

### PR DESCRIPTION
**Pull request by lrytz**
The method `missingParams` which returns undefined parameters of
a given invocation expression still assumed that named arguments can
only appear after positional ones.

### CODE REVIEW

**Comment by lrytz** [comment]
review by @som-snytt, maybe?

**Comment by som-snytt** [comment]
@lrytz Sure!  I'd forgotten that thread.

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala [hunk 1]
```diff
@@ -379,18 +379,37 @@ trait NamesDefaults { self: Analyzer =>
...
 
+  /**
+   * Returns the parameter symbols of an invocation expression that are not defined by the list
+   * of arguments.
+   *
+   * @param args    The list of arguments
+   * @param params  The list of parameter sybols of the invoked method
+   * @param argName A function that extracts the name of an argument expression, if it is a named argument.
+   */
   def missingParams[T](args: List[T], params: List[Symbol], argName: T => Option[Name] = nameOfNamedArg _): (List[Symbol], Boolean) = {
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: The default argName function is kind of weird.  It has to take an arbitrary T, i.e., Any, so it's odd to supply a default that is not just a backup for the most general, useless case.  (This is existing code.)

I did see that the T param is for Trees or Types.

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala [hunk 2]
```diff
@@ -379,18 +379,37 @@ trait NamesDefaults { self: Analyzer =>
...
-    val namedArgs = args.dropWhile(arg => {
-      val n = argName(arg)
-      n.isEmpty || params.forall(p => p.name != n.get)
-    })
-    val namedParams = params.drop(args.length - namedArgs.length)
+    // The argument list contains first a mix of positional args and named args that are on the
+    // right parameter position, and then a number or named args on different positions.
+
+    // collect all named arguments whose position does not match the parameter they define
+    val namedArgsOnChangedPosition = args.zip(params).dropWhile({
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: I would drop the extra paren for style points, `.dropWhile { etc }`.

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala [hunk 3]
```diff
@@ -379,18 +379,37 @@ trait NamesDefaults { self: Analyzer =>
...
+      case (arg, param) =>
+        val n = argName(arg)
+        // drop the argument if
+        //  - it's not named, or
+        //  - it's named, but defines the parameter on its current position, or
+        //  - it's named, but none of the parameter names matches (treated as a positional argument, an assignment expression)
+        n.isEmpty || n.get == param.name || params.forall(p => p.name != n.get)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: OK, got it. Do we prefer `params.forall(_.name != n.get)`?

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala [hunk 4]
```diff
@@ -379,18 +379,37 @@ trait NamesDefaults { self: Analyzer =>
...
+    }).map(_._1)
+
+    val paramsWithoutPositionalArg = params.drop(args.length - namedArgsOnChangedPosition.length)
+
     // missing: keep those with a name which doesn't exist in namedArgs
-    val missingParams = namedParams.filter(p => namedArgs.forall(arg => {
+    val missingParams = paramsWithoutPositionalArg.filter(p => namedArgsOnChangedPosition.forall(arg => {
       val n = argName(arg)
       n.isEmpty || n.get != p.name
     }))
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: Syntax: `namedArgsOCP.forall { arg => val n = etc ; n.isEmpty || etc }`.

The brace means you don't need your other brace, IIRC, where IIRC = I didn't try it.

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala [hunk 5]
```diff
@@ -379,18 +379,37 @@ trait NamesDefaults { self: Analyzer =>
...
+        // drop the argument if
+        //  - it's not named, or
+        //  - it's named, but defines the parameter on its current position, or
+        //  - it's named, but none of the parameter names matches (treated as a positional argument, an assignment expression)
+        n.isEmpty || n.get == param.name || params.forall(p => p.name != n.get)
+    }).map(_._1)
+
+    val paramsWithoutPositionalArg = params.drop(args.length - namedArgsOnChangedPosition.length)
+
     // missing: keep those with a name which doesn't exist in namedArgs
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: comment s/namedArgs/psWPA/

Does Scala IDE refactor names in comments?

**Comment by som-snytt** [comment]
LGTM.  I don't know why that's so tricky.  It should be "obviously LGTM".

I forgot to see what became of the original error message.  Can it still be emitted?  If so, the ML thread had a rewording.  But probably it doesn't happen anymore.  Maybe that should be an assert?

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala (hunk 1 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: Agree, I pushed it to the call-site.

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala (hunk 2 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: 1up

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala (hunk 3 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: 2up!

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala (hunk 4 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: Also true!

File: src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala (hunk 5 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala]: I didn't use the IDE here..

**Comment by lrytz** [comment]
The example compiles now, it was a bug that it didn't. So the confusing error message is gone.

Thanks for the review!

**Pull request merged**